### PR TITLE
Murisi/strategy constructors

### PR DESCRIPTION
**Pull request by murisi**
## Describe your changes
Updated the test vector generation code to make it easier to produce complete test vectors. More specifically, the changes are as follows:
* Test vector generation is now done offline using a Cargo `[[example]]` so that those needing them do not need to start a server to produce them
* Vector generation is now randomly produced with `proptest` strategies in order to more comprehensively cover the full domain of transactions
  * This approach has the additional benefit that these strategies can also be reused in other `proptest` tests
* Vector generation is now done in Rust in order to take advantage of Rust's type-checking and to ensure that the code is kept up to date
* The old approach to generating test vectors using shell scripts and hooks in the transaction generation and signing process have been removed as they are now redundant

## Indicate on which release or other PRs this topic is based on
#2257 

## Checklist before merging to `draft`
- [x] I have added a changelog
- [x] Git history is in acceptable state

### CODE REVIEW

File: core/src/types/hash.rs [hunk 1]
```diff
@@ -146,14 +146,17 @@ impl From<Hash> for crate::tendermint::Hash {
...
+/// Tests and strategies for hashes
+pub mod tests {
     use proptest::prelude::*;
+    #[cfg(test)]
     use proptest::string::{string_regex, RegexGeneratorStrategy};
 
     use super::*;
 
     /// Returns a proptest strategy that yields hex encoded hashes.
+    #[cfg(test)]
```
Comment (sug0) [inline core/src/types/hash.rs:159]: odd looking hack, since the module is already flagged as a test module

**Review by sug0 — COMMENTED** [review]
couple of comments

File: core/src/ledger/ibc/mod.rs [hunk 2]
```diff
@@ -296,3 +296,179 @@ pub fn received_ibc_token(
...
+    prop_compose! {
+        /// Generate an arbitrary IBC timestamp
+        pub fn arb_ibc_timestamp()(nanoseconds: u64) -> Timestamp {
+            Timestamp::from_nanoseconds(nanoseconds).expect("generated invalid IBC timestamp")
+        }
+    }
+
+    prop_compose! {
+        /// Generate an arbitrary IBC memo
+        pub fn arb_ibc_memo()(memo in "[a-zA-Z0-9_]*") -> Memo {
```
[the remark this answers is not in the document]
Comment (tzemanovic) [reply core/src/ledger/ibc/mod.rs:363]: `\w` produces utf8 chars - let's keep this pattern

File: core/src/types/eth_bridge_pool.rs [hunk 3]
```diff
@@ -302,6 +302,77 @@ pub struct GasFee {
...
+        }
+    }
+
+    prop_compose! {
+        /// Generate an arbitrary transfer to Ethereum
+        pub fn arb_transfer_to_ethereum()(
+            kind in arb_transfer_to_ethereum_kind(),
+            asset in arb_eth_address(),
+            recipient in arb_eth_address(),
+            sender in arb_address(),
```
[the remark this answers is not in the document]
Comment (tzemanovic) [reply core/src/types/eth_bridge_pool.rs:362]: `arb_established_address` for tokens

File: core/src/types/token.rs [hunk 4]
```diff
@@ -1303,9 +1303,51 @@ pub enum TransferError {
...
+            DenominatedAmount::new(amount, denom)
+        }
+    }
+
+    prop_compose! {
+        /// Generate a transfer
+        pub fn arb_transfer()(
+            source in arb_address(),
+            target in arb_address(),
+            token in arb_address(),
```
Comment (tzemanovic) [inline core/src/types/token.rs:1337]: also for these

**Review by tzemanovic — APPROVED** [review]

File: core/src/types/eth_bridge_pool.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (tzemanovic) [reply core/src/types/eth_bridge_pool.rs]: done in [c9092ca](https://github.com/anoma/namada/pull/2259/commits/c9092ca4eaf499e76efb53e916706fa6066857c9)

File: core/src/types/eth_bridge_pool.rs [hunk 5]
```diff
@@ -302,6 +302,77 @@ pub struct GasFee {
...
+            }
+        }
+    }
+
+    prop_compose! {
+        /// Generate an arbitrary Ethereum gas fee
+        pub fn arb_gas_fee()(
+            amount in arb_amount(),
+            payer in arb_address(),
+            token in arb_address(),
```
[the remark this answers is not in the document]
Comment (tzemanovic) [reply core/src/types/eth_bridge_pool.rs]: [c9092ca](https://github.com/anoma/namada/pull/2259/commits/c9092ca4eaf499e76efb53e916706fa6066857c9)

File: core/src/types/hash.rs (hunk 1 above)
Comment (tzemanovic) [reply core/src/types/hash.rs]: I've refactored the testing modules a bit in here [6936ac2](https://github.com/anoma/namada/pull/2259/commits/6936ac2c3c564ac55eb6c45256fb313068752851) - the strategies and test helpers should be separate sub-module (`mod testing`) and actual tests in `mod test(s)`

Comment (tzemanovic) [reply core/src/types/hash.rs]: so that "testing" can be re-used in crates that depend on these

**Review by sug0 — APPROVED** [review]

**Pull request merged**